photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala

```diff
@@ -0,0 +1,252 @@
+
+    if (downSamplingRate > 0D && downSamplingRate < 1D) {
+      taskType match {
+        case TaskType.LOGISTIC_REGRESSION | TaskType.SMOOTHED_HINGE_LOSS_LINEAR_SVM =>
+          Some(new BinaryClassificationDownSampler(downSamplingRate))
+
+        case TaskType.LINEAR_REGRESSION | TaskType.POISSON_REGRESSION =>
+          Some(new DefaultDownSampler(downSamplingRate))
+      }
+    } else {
+      None
+    }
+
+  private def selectRandomLatentObjectiveFunction(
+      randomEffectOptimizationConfiguration: GLMOptimizationConfiguration,
+      latentFactorOptimizationConfiguration: GLMOptimizationConfiguration):
+      (SingleNodeObjectiveFunction, DistributedObjectiveFunction) = {
+
+    val lossFunction: LossFunction  = SingleNodeGLMLossFunction(randomEffectOptimizationConfiguration)
+    val distLossFunction: DistributedLossFunction =
+      DistributedGLMLossFunction(sc, latentFactorOptimizationConfiguration, DEFAULT_TREE_AGGREGATE_DEPTH)
+
+    taskType match {
+      case TaskType.LOGISTIC_REGRESSION => (lossFunction(LogisticLossFunction), distLossFunction(LogisticLossFunction))
+      case TaskType.LINEAR_REGRESSION => (lossFunction(SquaredLossFunction), distLossFunction(SquaredLossFunction))
+      case TaskType.POISSON_REGRESSION => (lossFunction(PoissonLossFunction), distLossFunction(PoissonLossFunction))
+      case TaskType.SMOOTHED_HINGE_LOSS_LINEAR_SVM =>
+        (SingleNodeSmoothedHingeLossFunction(randomEffectOptimizationConfiguration),
+          DistributedSmoothedHingeLossFunction(sc, latentFactorOptimizationConfiguration, DEFAULT_TREE_AGGREGATE_DEPTH))
+    }
+  }
+
+  private def selectSingleNodeLossFunction(configuration: GLMOptimizationConfiguration): SingleNodeObjectiveFunction = {
+
+    val lossFunction: LossFunction = SingleNodeGLMLossFunction(configuration)
+
+    taskType match {
+      case TaskType.LOGISTIC_REGRESSION => lossFunction(LogisticLossFunction)
+      case TaskType.LINEAR_REGRESSION => lossFunction(SquaredLossFunction)
+      case TaskType.POISSON_REGRESSION => lossFunction(PoissonLossFunction)
+      case TaskType.SMOOTHED_HINGE_LOSS_LINEAR_SVM => SingleNodeSmoothedHingeLossFunction(configuration)
+    }
+  }
+
+  private def selectDistributedLossFunction(
+      configuration: GLMOptimizationConfiguration,
+      treeAggregateDepth: Int): DistributedObjectiveFunction = {
+
+    val distLossFunction: DistributedLossFunction = DistributedGLMLossFunction(sc, configuration, treeAggregateDepth)
+
+    taskType match {
+      case TaskType.LOGISTIC_REGRESSION => distLossFunction(LogisticLossFunction)
+      case TaskType.LINEAR_REGRESSION => distLossFunction(SquaredLossFunction)
+      case TaskType.POISSON_REGRESSION => distLossFunction(PoissonLossFunction)
+      case TaskType.SMOOTHED_HINGE_LOSS_LINEAR_SVM =>
+        DistributedSmoothedHingeLossFunction(sc, configuration, treeAggregateDepth)
+    }
+  }
+}
+
+object GameEstimator {
+  val FIXED_EFFECT_FEATURE_THRESHOLD = 200000
```

Comment:
I'm open to suggestions for where to move these constants - though I don't think the parameters is a good place since these settings aren't meant to be tuned (unless we want to make them be configurable, in which case ...)